logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -306,23 +306,35 @@ private boolean isOutput(final Vertex vertex) {
 
     public final class CompiledOrderedExecution extends CompiledExecution {
 
+        @SuppressWarnings({"unchecked"})  private final RubyArray<RubyEvent> EMPTY_ARRAY = RubyUtil.RUBY.newEmptyArray();
+
         @Override
         public void compute(final QueueBatch batch, final boolean flush, final boolean shutdown) {
            compute(batch.events(), flush, shutdown);
         }
 
         @Override
         public void compute(final Collection<RubyEvent> batch, final boolean flush, final boolean shutdown) {
-            @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
-            // send batch one-by-one as single-element batches down the filters
-            @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> filterBatch = RubyUtil.RUBY.newArray(1);
-            for (final RubyEvent e : batch) {
-                filterBatch.set(0, e);
-                final Collection<RubyEvent> result = compiledFilters.compute(filterBatch, flush, shutdown);
-                copyNonCancelledEvents(result, outputBatch);
-                compiledFilters.clear();
+            if (!batch.isEmpty()) {
+                @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+                @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> filterBatch = RubyUtil.RUBY.newArray(1);
+                // send batch one-by-one as single-element batches down the filters
+                for (final RubyEvent e : batch) {
+                    filterBatch.set(0, e);
+                    _compute(filterBatch, outputBatch, flush, shutdown);
+                }
+                compiledOutputs.compute(outputBatch, flush, shutdown);
+            } else if (flush) {
+                @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+                _compute(EMPTY_ARRAY, outputBatch, true, shutdown);
+                compiledOutputs.compute(outputBatch, true, shutdown);
```

Comment:
```suggestion
            @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
            if (batch.isEmpty() && flush) {
                _compute(EMPTY_ARRAY, outputBatch, true, shutdown);
            } else {
                @SuppressWarnings({"unchecked"}) final RubyArray<RubyEvent> filterBatch = RubyUtil.RUBY.newArray(1);
                // send batch one-by-one as single-element batches down the filters
                for (final RubyEvent e : batch) {
                    filterBatch.set(0, e);
                    _compute(filterBatch, outputBatch, flush, shutdown);
                }
            }
            compiledOutputs.compute(outputBatch, flush, shutdown);
```

Reply:
Do think this could be more readable, extracting the head and tail common part of condition body?

Reply:
This is what I had in the first place but realized that the `else` clause will catch the condition of `batch.isEmpty() == true && flush == false` which we don't really want so that's why I shuffled the condition to make that that upon an empty batch and no flush required, nothing happens.

Reply:
the `else` should catch `batch.isEmpty()==false || flush == false`

Reply:
but is LGTM, go with the merge for me